bindings/rust/s2n-tls/src/callbacks.rs

```diff
@@ -57,91 +57,180 @@ where
-/// |   +-> return Poll::Ready                  (Rust)
-/// |   s2n_negotiate                           (C)
-/// |                          
+/// Connection::poll_negotiate                                    (Rust)
+/// |   poll_client_hello_callback                                (Rust)
+/// |   |   ConnectionFuture::poll                                (Rust)
+/// |   |   +-> return Poll::Ready                                (Rust)
+/// |   +-> return Poll::Ready                                    (Rust)
+/// |   s2n_negotiate                                             (C)
+/// |
 /// v   ...handshake continues.
 ///
 /// Note that "s2n_client_hello_cb" is only called once.
 /// After the initial call, the retries are handled by the Rust bindings.
 /// s2n_negotiate is not called again until the callback completes.
 ///
-pub(crate) fn trigger_async_callback<T: 'static + AsyncCallback>(
-    mut callback: T,
-    conn: &mut Connection,
-) -> CallbackResult {
+pub(crate) fn trigger_async_client_hello_callback(conn: &mut Connection) -> CallbackResult {
     // Try once first.
-    match callback.poll(conn) {
+    match poll_client_hello_callback(conn, None) {
         // If callback completes, no need for retry.
         Poll::Ready(r) => r.into(),
         // If callback doesn't complete, prepare connection for retry.
-        Poll::Pending => {
-            conn.set_pending_callback(Some(Box::new(callback)));
-            CallbackResult::Success
-        }
+        Poll::Pending => CallbackResult::Success,
     }
 }
 
-/// An asynchronous adapter for an s2n-tls callback.
+/// The Future associated with the async connection callback.
 ///
-/// This trait must be implemented for any asynchronous callbacks
-/// supported by the bindings.
+/// The calling application can provide an instance of [`ConnectionFuture`]
+/// when implementing an async callback, eg. [`ClientHelloCallback`], if it wants
+/// to run an asynchronous operation (disk read, network call). The application
+/// can return an error ([`Err(error::Error::application())`]), to indicate
+/// connection failure.
 ///
-/// Implementations should capture any arguments passed to the callback
-/// so that they will be available for every call to [`AsyncCallback::poll`].
-/// [`AsyncCallback::poll`] should handle retrieving the callback from
-/// the connection or config, passing it the stored arguments,
-/// and calling any "mark done" style methods necessary to unblock the
-/// connection once the callback has succeeded.
-pub(crate) trait AsyncCallback {
-    fn poll(&mut self, conn: &mut Connection) -> Poll<Result<(), Error>>;
+/// [`ConfigResolver`] should be used if the application wants to set a new
+/// [`Config`] on the connection.
+pub trait ConnectionFuture {
+    fn poll(
+        self: Pin<&mut Self>,
+        connection: &mut Connection,
+        ctx: &mut core::task::Context,
+    ) -> Poll<Result<(), Error>>;
+}
+
+pin_project! {
```

Comment:
We need this because the [Future::poll](https://doc.rust-lang.org/std/future/trait.Future.html#tymethod.poll) method expects a `Pin<&mut Self>`.

The [docs on pin](https://doc.rust-lang.org/std/pin/index.html#projections-and-structural-pinning) has more info on projection.